File: esperanto-ios/Esperanto/JSContext.swift
//
//  JSContext.swift
//  Esperanto
//
//  Created by Alastair Coote on 8/4/20.
//  Copyright © 2020 Alastair Coote. All rights reserved.
//

import Foundation
import Esperanto.Esperanto_Private

public class JSContext {
    let ctx: OpaquePointer
    public init() {
        self.ctx = jscontext_new()
    }

    public func evaluate(script: String) -> JSValue {
        let val_ptr = jscontext_evaluate(self.ctx, script)
        return JSValue(ptr: val_ptr!)
    }

    public func compile(script:String) -> UnsafeMutablePointer<CompiledCode> {
        let compiled = jscontext_compile_string(self.ctx, script)!
        return compiled
//        let buffer = UnsafeBufferPointer(start: compiled.pointee.bytes, count: Int(compiled.pointee.len));
//        return Array(buffer)
    }

    public func evaluate(compiledCode: UnsafeMutablePointer<CompiledCode>) -> JSValue {
        let val_ptr = jscontext_eval_compiled(self.ctx, compiledCode)
        return JSValue(ptr: val_ptr!)
    }

    deinit {
        jscontext_free(self.ctx)
    }
}

File: esperanto-android/esperanto/src/main/java/org/esperanto/esperanto/JSContext.kt
package org.esperanto.esperanto

private object JSContextPrivate {
    @JvmStatic external fun new() : Long;
    @JvmStatic external fun evaluate(ctx_ptr: Long, script:String) : Long;

    init {
        System.loadLibrary("esperanto")
    }
}

class JSContext {

    private var ptr:Long = JSContextPrivate.new()

    fun evaluate(script:String) : JSValue {
        val result_ptr = JSContextPrivate.evaluate(ptr, script)
        return JSValue(result_ptr)
    }


    external fun instanceTest() : String;

}